fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/InComparisonToExplodeRule.java

```diff
@@ -149,7 +149,9 @@ public void onMatch(@Nonnull final CascadesRuleCall call) {
                 final var comparison = valuePredicate.getComparison();
                 Verify.verify(comparison.getType() == Comparisons.Type.IN);
                 final ExplodeExpression explodeExpression;
-                if (comparison instanceof Comparisons.ListComparison) {
+                if (comparison instanceof Comparisons.ValueComparison && ((Comparisons.ValueComparison)comparison).getComparandValue().getResultType().getTypeCode().equals(Type.TypeCode.ARRAY)) {
```

Comment:
This is not exactly a nice way to write this. I have run into this as well. Can you add methods on `Type` for the generic types like record and array at least: `isArray()` and `isRecord()` as well as `Type.Record narrowRecord()` and `Type.Array narrowArray()` that do the appropriate things for those. Maybe enums as well...

Reply:
Another thought here: If you already have a `comparison.getType() == Comparisons.Type.IN`, can't you assume the result type is of type array? Seems like a result type of the value of type `INT` or something is an error to start with.

Reply:
refactored.